### DISCUSSION TITLE
ErrorList

### DISCUSSION THREAD

**Discussion by xavierjohn** [vkhorikov/CSharpFunctionalExtensions#462] [Show and tell]
I created a nuget package that will help return errors as a collection. 
I have also posted a sample code of how I think fluent validation can be leveraged in Domain which is different from the idea presented in the Pluralsite course.

Let me know your thoughts.
https://www.nuget.org/packages/CSharpFunctionalExtensions.Errors/0.0.0.1-alpha

```csharp
public class User : Entity<Guid>
{
    public string FirstName { get; }
    public string LastName { get; }
    public string Email { get; }
    public string Password { get; }

    public static Result<User, ErrorList> Create(string firstName, string lastName, string email, string password)
    {
        var user = new User(firstName, lastName, email, password);
        return s_validator.Validate(user).ToResult(user);
    }


    private User(string firstName, string lastName, string email, string password)
    {
        FirstName = firstName;
        LastName = lastName;
        Email = email;
        Password = password;
    }

    static readonly InlineValidator<User> s_validator = new()
    {
        v => v.RuleFor(x => x.FirstName).NotEmpty(),
        v => v.RuleFor(x => x.LastName).NotEmpty(),
        v => v.RuleFor(x => x.Email).NotEmpty().EmailAddress(),
        v => v.RuleFor(x => x.Password).NotEmpty()
    };
}
```

**Comment by vkhorikov** [comment]
Thank you, this is going to be a good reference point. I'll let you know once I start working on this, so that you can review the implementation.

**Comment by xavierjohn** [comment]
I also wrote a base controller
https://www.nuget.org/packages/CSharpFunctionalExtensions.Asp/0.0.0.1-alpha

The library converts CSharpFunctionalExtensions.Result to HTTP errors but using the base controller
`CSharpFunctionalBase` and calling the method `MapToOkObjectResult`

Run the sample program to try it out.

Example:

```csharp
   public ActionResult<AuthenticationResult> Register(RegisterRequest request) =>
        Domain.User.Create(request.firstName, request.lastName, request.email, request.password)
             .Bind(user => Result.Success<AuthenticationResult, ErrorList>(new AuthenticationResult(user, "token")))
             .Finally(authResult => MapToOkObjectResult(authResult));
```

If a validation failure occurs, it will return a response like

```
HTTP/1.1 400 Bad Request
Connection: close
Content-Type: application/problem+json; charset=utf-8
Date: Mon, 14 Nov 2022 20:29:19 GMT
Server: Kestrel
Transfer-Encoding: chunked
```

```json
{
  "type": "https://tools.ietf.org/html/rfc7231#section-6.5.1",
  "title": "One or more validation errors occurred.",
  "status": 400,
  "traceId": "00-19b910c84bc8651a1c6d26e9ee640813-59f0f7943ce1d3ab-00",
  "errors": {
    "lastName": [
      "The lastName field is required."
    ],
    "firstName": [
      "The firstName field is required."
    ]
  }
}
```

**Comment by xavierjohn** [comment]
I updated the package so there is no longer the need to derive from my base controller, instead use the extension method `ToActionResult`

```csharp
   public ActionResult<AuthenticationResult> Register(RegisterRequest request) =>
        Domain.User.Create(request.firstName, request.lastName, request.email, request.password)
             .Bind(user => Result.Success<AuthenticationResult, ErrorList>(new AuthenticationResult(user, "token")))
             .ToActionResult(this);
```

https://www.nuget.org/packages/CSharpFunctionalExtensions.Asp/0.1.0-alpha.5

**Comment by dataphysix** [comment]
I have dealt with the same issue a year ago, but wanted a single Error class to handle this transparently. Basically a quick and dirty variation on what @xavierjohn provided :
[https://github.com/dataphysix/CSharpFunctionalExtensions.Error/blob/main/Errors.cs](url)

@vkhorikov Any eta on v3 ?  Was it already decided if the Error class will become a part of the library ?

**Reply by vkhorikov** [reply]
I'm trying to start this release in earnest but keep postponing it due to lack of time. But yes, Error class is going to be part of v3, it will contain Code and Message properties (you can find the details here https://github.com/vkhorikov/CSharpFunctionalExtensions/issues/321 ).

**Comment by xavierjohn** [comment]
I took a different approach of only using a collection with AggregatedError and ValidationError.
https://github.com/xavierjohn/FunctionalDDD/tree/main/RailwayOrientedProgramming/src/Errors/Types

The v1 version of the library is out https://www.nuget.org/packages/FunctionalDDD.RailwayOrientedProgramming/1.0.1 since the first service using it has shipped to production.
